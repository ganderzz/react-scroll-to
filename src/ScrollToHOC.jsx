import React from "react";
import getDisplayName from "./utilities/getDisplayName";
import ScrollTo from "./ScrollTo";

/**
 * Higher Order Component version of the ScrollTo.
 * Injects a prop named scroll that is a function that
 * takes an (x, y) coordinate to scroll to. [ie. props.scrollTo(0, 500)]
 */
function ScrollToHOC(Component) {
  const WrappedComponent = props => (
    <ScrollTo>
      {scrollProps => <Component {...props} {...scrollProps} />}
    </ScrollTo>
  );
  WrappedComponent.displayName = `WithScrollToHOC(${getDisplayName(
    Component
  )})`;

  return WrappedComponent;
}

export default ScrollToHOC;
