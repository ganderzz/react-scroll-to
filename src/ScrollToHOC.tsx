import * as React from "react";
import { ScrollTo } from "./ScrollTo";

function getDisplayName(Component) {
  const { displayName, name } = Component;

  if (displayName || name) {
    return displayName || name;
  }
  if (typeof Component === "string" && Component.length > 0) {
    return Component;
  }

  return "ScrollTo/Unknown";
}

/**
 * Higher Order Component version of the ScrollTo.
 * Injects a prop named scroll that is a function that
 * takes an (x, y) coordinate to scroll to. [ie. props.scrollTo(0, 500)]
 */
function ScrollToHOC(Component) {
  const WrappedComponent = (props) => (
    <ScrollTo>
      {(scrollProps) => <Component {...props} {...scrollProps} />}
    </ScrollTo>
  );

  WrappedComponent.displayName = `WithScrollToHOC(${getDisplayName(
    Component
  )})`;

  return WrappedComponent;
}

export { ScrollToHOC };
