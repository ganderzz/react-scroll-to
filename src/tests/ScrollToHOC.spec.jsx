import React from "react";
import { render, fireEvent, cleanup } from "react-testing-library";
import ScrollToHOC from "../ScrollToHOC";

afterAll(cleanup);

beforeEach(() => {
  window.scrollTo = jest.fn();
});

describe("Test HOC.", () => {
  it("Should render the functional children.", () => {
    const TestComponent = () => <div>test</div>;
    TestComponent.displayName = "test";

    const WrappedComponent = ScrollToHOC(TestComponent);
    const { container } = render(<WrappedComponent />);

    expect(container).toMatchSnapshot();
  });

  it("Should call window.scroll.", () => {
    const TestComponent = props => (
      <button onClick={() => props.scrollTo({ x: 100, y: 200 })}>mybtn</button>
    );
    TestComponent.displayName = "test";

    const WrappedComponent = ScrollToHOC(TestComponent);
    const { getByText } = render(<WrappedComponent />);

    fireEvent.click(getByText("mybtn"));

    expect(window.scrollTo).toHaveBeenCalledTimes(1);
    expect(window.scrollTo.mock.calls[0]).toEqual([
      {
        left: 100,
        top: 200,
        behavior: "auto"
      }
    ]);
  });

  it("Should call scrollById.", () => {
    const WrappedComponent = ScrollToHOC(props => {
      return (
        <div>
          <div id="foo" />
          <button onClick={() => props.scrollTo({ id: "foo", x: 100, y: 200 })}>
            test
          </button>
        </div>
      );
    });

    const { getByText, container } = render(<WrappedComponent />);

    fireEvent.click(getByText("test"));

    expect(container).toMatchSnapshot();
  });
});
