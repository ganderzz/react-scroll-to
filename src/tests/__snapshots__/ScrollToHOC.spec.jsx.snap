// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HOC. Should call scrollById. 1`] = `
<div>
  <div>
    <div
      id="foo"
    />
    <button>
      test
    </button>
  </div>
</div>
`;

exports[`Test HOC. Should render the functional children. 1`] = `
<div>
  <div>
    test
  </div>
</div>
`;
